LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDeviceClientConnectionManager.cs

```diff
@@ -1,162 +1,399 @@
-            lock (this.managedConnections)
+            public Task<Twin> GetTwinAsync(CancellationToken cancellationToken) =>
+                InvokeExclusivelyAsync(client => client.GetTwinAsync(cancellationToken));
+
+            public Task<bool> SendEventAsync(LoRaDeviceTelemetry telemetry, Dictionary<string, string> properties) =>
+                InvokeExclusivelyAsync(client => client.SendEventAsync(telemetry, properties));
+
+            public Task<bool> UpdateReportedPropertiesAsync(TwinCollection reportedProperties, CancellationToken cancellationToken) =>
+                InvokeExclusivelyAsync(client => client.UpdateReportedPropertiesAsync(reportedProperties, cancellationToken));
+
+            public Task<Message> ReceiveAsync(TimeSpan timeout) =>
+                InvokeExclusivelyAsync(client => client.ReceiveAsync(timeout));
+
+            public Task<bool> CompleteAsync(Message cloudToDeviceMessage) =>
+                InvokeExclusivelyAsync(client => client.CompleteAsync(cloudToDeviceMessage));
+
+            public Task<bool> AbandonAsync(Message cloudToDeviceMessage) =>
+                InvokeExclusivelyAsync(client => client.AbandonAsync(cloudToDeviceMessage));
+
+            public Task<bool> RejectAsync(Message cloudToDeviceMessage) =>
+                InvokeExclusivelyAsync(client => client.RejectAsync(cloudToDeviceMessage));
+
+            public bool EnsureConnected()
             {
-                if (this.managedConnections.ContainsKey(key))
+                var connected = this.client.EnsureConnected();
+
+                if (this.logger?.IsEnabled(LogLevel.Debug) ?? false)
                 {
-                    throw new InvalidOperationException($"Connection already registered for device {loRaDevice.DevEUI}");
+                    const string message = $"{nameof(EnsureConnected)} = {{Connected}}";
+                    this.logger.LogDebug(message, connected);
                 }
 
-                this.managedConnections[key] = new ManagedConnection(loRaDevice, loraDeviceClient);
+                if (connected)
+                    EnsureConnectedSucceeded?.Invoke(this, EventArgs.Empty);
+
+                return connected;
             }
-        }
 
-        public void Release(LoRaDevice loRaDevice)
-        {
-            _ = loRaDevice ?? throw new ArgumentNullException(nameof(loRaDevice));
+            public Task DisconnectAsync() =>
+                DisconnectAsync(deferred: false);
+
+            private enum DisconnectionResult { Disconnected, Deferred, NotDisconnected }
 
-            if (this.managedConnections.TryRemove(GetConnectionCacheKey(loRaDevice.DevEUI), out var removedItem))
+            private Task<DisconnectionResult> DisconnectAsync(bool deferred) =>
+                InvokeExclusivelyAsync(doesNotRequireOpenConnection: true, async client =>
+                {
+                    DisconnectionResult result;
+
+                    if (Interlocked.Add(ref this.activities, 0) > 0)
```

Comment:
not sure what we are trying to do here. The BeginDeviceClientConnectionActivity can be called any time. Why not just compare to 0 here instead? If you are racing, we have the race anyway and even if it is 0 now, it can be 1 on the next line, causing the disconnect to happen. 

Operations will eventually re-open the connection any they are synchronized, so not a big deal but the validation with 0 seems not to require an Interlocked.Add

Reply:
As discussed. I leave it up to you what you want to do with this. I feel it's adding a false sense of isolation. The next line the state can be changed no matter what you do on 264. The decision is always (potentially) outdated and different from the current state.